src/charm.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+
+"""Charm the calibre-web application."""
+
+import logging
+import typing
+from pathlib import Path
+from typing import cast
+
+import ops
+
+CONTAINER_NAME = 'calibre-web'
+SERVICE_NAME = 'calibre-web'
+STORAGE_NAME = 'books'
+LIBRARY_WRITE_ACTION = 'library-write'
+LIBRARY_WRITE_CONFIG = 'library-write'
+LIBRARY_INFO_ACTION = 'library-info'
+LIBRARY_INFO_FORMAT_PARAM = 'format'
+
+LibraryInfoFormat: typing.TypeAlias = typing.Literal['tree', 'ls-1']
+LIBRARY_INFO_FORMATS = typing.get_args(LibraryInfoFormat)
+
+LibraryWriteBehaviour: typing.TypeAlias = typing.Literal['skip', 'clean']
+LIBRARY_WRITE_BEHAVIOURS = typing.get_args(LibraryWriteBehaviour)
+
+
+CalibreWebLayerDict = typing.TypedDict(
+    'CalibreWebLayerDict',
+    {
+        'summary': str,
+        'description': str,
+        'services': dict[str, ops.pebble.ServiceDict],
+    },
+    total=True,
+)
+
+
+class CalibreWebCharm(ops.CharmBase):
+    """Charm the calibre-web application."""
+
+    def __init__(self, framework: ops.Framework) -> None:
+        super().__init__(framework)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        framework.observe(self.on[CONTAINER_NAME].pebble_ready, self._on_pebble_ready)
+        framework.observe(self.on[STORAGE_NAME].storage_attached, self._on_storage_attached)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
```

Comment:
This just seems to me like the right way to handle string values coming from `charmcraft.yaml`. What do you think?

Benefits I found while working on this charm were ease of renaming values in both `charmcraft.yaml` and `src/charm.py`, and not having to type the literal string values in my tests (and keep them up to date too).

I'm also hopeful that I can make changes to `ops` that will let the `on[NAME].event` form provide exact event types, which doesn't seem like it would be possible with `on.name_event`.

Reply:
I actually feel like it would be kind of neat if ops provided a nice way to pull the values directly from the `charmcraft.yaml` and really have a single source of truth, but that's probably a lot of over engineering for what is probably a complete non-issue in practice.

Reply:
Or if charmcraft could pull your actions and config from the charm code to populate actions.yaml and config.yaml :)

This ties a bit into the discussion and work we have planned around config/action/relation types, having a useful single source of truth for those, etc. Before we get to that we can hopefully at least lint that there aren't mismatches.

Reply:
> Or if charmcraft could pull your actions and config from the charm code to populate actions.yaml and config.yaml :)

That would be cool!

> In terms of on[NAME_SET_TO_FOO] vs on['foo'], I guess it does reduce repetition (particularly in tests?) and so make it easier to rename, but the plain string looks cleaner to me - maybe I'm just used to it?

Definitely the fact that I went with `on[LIBRARY_INFO_ACTION].action` rather than `on[LIBRARY_INFO].action` makes this uglier than perhaps it needs to be.